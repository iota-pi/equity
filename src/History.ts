export const MAX_GROUPING_TIME = 1200;
type NumberGroup = number[];

export class PlayerHistory {
  private data: NumberGroup[];
  private players: number;
  private lastEdit: number;

  constructor (players: number) {
    this.data = [[]];
    this.players = players;
    this.lastEdit = 0;
  }

  get numberOfGroups () {
    return this.data.length;
  }

  get numberOfCalls () {
    return this.data.reduce((sum: number, group: NumberGroup) => sum + group.length, 0);
  }

  get allCalls () {
    return this.data.reduce((calls: number[], group: NumberGroup) => calls.concat(group), []);
  }

  get callCounts (): Map<number, number> {
    const counts = new Map();
    for (let i = 0; i < this.players; ++i) {
      counts.set(i, 0);
    }

    for (let group of this.data) {
      for (let x of group) {
        counts.set(x, counts.get(x) + 1);
      }
    }

    return counts;
  }

  get numberOfPlayers () {
    return this.players;
  }

  get rawData () {
    return this.data;
  }

  add (x?: number) {
    // Get random value for x, or ensure it is in appropriate range (0 <= x < players)
    x = x === undefined ? this.randomPlayer() : Math.abs(Math.floor(x)) % this.players;

    // Start a new group if current group is not empty and enough time has elapsed
    if (this.timeElapsed > MAX_GROUPING_TIME) {
      this.newGroup();
    }

    // Start a new group if this value already exists in the current group
    if (this.lastGroup.includes(x)) {
      this.newGroup();
    }

    this.lastGroup.push(x);
    this.setTime();

    return x;
  }

  newGroup () {
    // Only create a new group if current last group is not empty
    if (this.lastGroup.length > 0) {
      this.data.push([]);
    }
  }

  undo () {
    if (this.lastGroup.length > 0) {
      this.lastGroup.pop();
    } else if (this.numberOfGroups > 1) {
      this.data.pop();
      this.lastGroup.pop();
    }
    this.setTime();
  }

  clear () {
    this.data = [[]];
    this.lastEdit = 0;
  }

  formatted () {
    const numberSort = (a: number, b: number) => +(a > b) - +(a < b);
    const formatted = this.data.map((group) => group.slice().sort(numberSort));
    return formatted.filter(group => group.length > 0);
  }

  load (data: NumberGroup[]) {
    this.data = data;
    return this;
  }

  private randomPlayer () {
    const counts = this.callCounts;
    const smallestCount = Math.min(...Array.from(counts.values()))

    // Get list of possible choices
    let choices = (new Array(this.players)).fill(0).map((_, i) => i);

    // Only pick one of the least called players
    choices = choices.filter(player => counts.get(player) === smallestCount);

    // Try to avoid picking one that's already been chosen in this group
    const notInGroup = choices.filter(player => !this.lastGroup.includes(player));
    if (notInGroup.length > 0) {
      choices = notInGroup;
    }

    // If there are still multiple options, don't call the same player as last time
    if (choices.length > 1) {
      const avoidCall = this.allCalls[this.numberOfCalls - 1];
      choices = choices.filter((player => player !== avoidCall));
    }

    const p = choices[Math.floor(Math.random() * choices.length)];

    return p;
  }

  private setTime () {
    this.lastEdit = (new Date()).getTime();
  }

  private get timeElapsed () {
    const now = (new Date()).getTime();
    return now - this.lastEdit;
  }

  private get lastGroup () {
    return this.data[this.data.length - 1];
  }
}
