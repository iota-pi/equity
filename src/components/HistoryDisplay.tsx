import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import { History } from '../History';
import withStyles, { WithStyles } from "@material-ui/core/styles/withStyles";
import createStyles from "@material-ui/core/styles/createStyles";
import { Theme } from "@material-ui/core/styles/createMuiTheme";
import Typography from '@material-ui/core/Typography';
import FlipMove from 'react-flip-move';

const styles = (theme: Theme) => createStyles({
  root: {
    position: 'absolute',
    left: -100,
    right: -100,
    top: 0,
    bottom: 0,
  },
  scrollable: {
    display: 'flex',
    flexDirection: 'column-reverse',
    overflowY: 'scroll',
    zIndex: 10,
    paddingLeft: 100 + theme.spacing.unit * 2,
    paddingRight: 100 + theme.spacing.unit * 2,
    height: '100%',
  },
  historyContainer: {
    display: 'flex',
    flexDirection: 'column-reverse',
    minHeight: `calc(${theme.typography.h1.fontSize} * 2)`,
    '&::before, &::after': {
      display: 'block',
      content: '""',
      position: 'absolute',
      left: 0,
      right: 0,
      zIndex: 100,
      height: theme.spacing.unit * 20,
    },
    '&::before': {
      top: 0,
      background: `linear-gradient(${theme.palette.background.default}, transparent)`,
    },
    '&::after': {
      bottom: 0,
      background: `linear-gradient(transparent, ${theme.palette.background.default})`,
    },
  },
  historyRow: {
    textAlign: 'center',
    width: '100%',
    marginBlock: 0,
    '&>span': {
      transition: 'all 0.3s ease',
    },
    [theme.breakpoints.only('xs')]: {
      fontSize: `calc(${theme.typography.h2.fontSize} * 0.85)`,
      '&:last-of-type': {
        '&>span': {
          color: theme.palette.primary.main,
          fontSize: `calc(${theme.typography.h1.fontSize})`,
        },
      },
    },
    [theme.breakpoints.up('sm')]: {
      fontSize: `calc(${theme.typography.h2.fontSize})`,
      '&:last-of-type': {
        '&>span': {
          color: theme.palette.primary.main,
          fontSize: `calc(${theme.typography.h1.fontSize} * 1.15)`,
        },
      },
    },
  },
  verticalSpacer: {
    height: theme.spacing.unit * 20,
  },
});

export interface Props extends WithStyles<typeof styles> {
  history: History,
  names: string[],
}

class HistoryDisplay extends Component<Props> {
  private scrollBarWidth: number = 0

  render() {
    const history = this.props.history.formatted();
    const classes = this.props.classes;

    return (
      <div className={classes.root}>
        <div
          className={classes.scrollable}
          ref="scrollable"
          style={{marginRight: -this.scrollBarWidth}}
        >
          <div className={classes.historyContainer}>
            <FlipMove
              enterAnimation={{
                from: { opacity: '0', transform: 'translateY(50px)' },
                to: { opacity: '1', transform: 'translateY(0px)' },
              }}
              leaveAnimation={{
                from: { opacity: '1', transform: 'translateY(0px)' },
                to: { opacity: '0', transform: 'translateY(50px)' },
              }}
              easing="ease-out"
              duration={300}
            >
              <div className={classes.verticalSpacer}></div>
              {history.map((numbers, i) => (
                <Typography
                  key={`historyRow-${i}`}
                  variant="h1"
                  align="center"
                  className={classes.historyRow}
                >
                  {this.renderPlayers(numbers)}
                </Typography>
              ))}
              <div className={classes.verticalSpacer}></div>
            </FlipMove>
          </div>
        </div>
      </div>
    )
  }

  private renderPlayers = (players: number[]) => {
    const toString = (id: number) => {
      return this.props.names[id] || (id + 1).toString()
    }

    return (
      <React.Fragment>
        {players.map((p, i) => (
          <React.Fragment key={`player-${i}`}>
            {i > 0 && <span>, </span>}
            <span>{toString(p)}</span>
          </React.Fragment>
        ))}
      </React.Fragment>
    )
  }

  componentDidMount () {
    // Get scrollbar width
    const el = ReactDOM.findDOMNode(this.refs.scrollable) as HTMLElement;
    this.scrollBarWidth = el.offsetWidth - el.clientWidth;
  }

  componentDidUpdate () {
    this.scrollToBottom();
  }

  private scrollToBottom = () => {
    const el = ReactDOM.findDOMNode(this.refs.scrollable) as HTMLElement;
    el.scroll({
      top: el.scrollHeight,
      behavior: 'smooth',
    });
  }
}

export default withStyles(styles)(HistoryDisplay);
