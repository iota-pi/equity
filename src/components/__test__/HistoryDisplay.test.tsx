import React from 'react';
import HistoryDisplay from '../HistoryDisplay';
import { shallow, mount, render } from 'enzyme';
import { PlayerHistory } from '../../History';

describe('<HistoryDisplay/>', () => {
  it('renders base state consistently', () => {
    const h = new PlayerHistory(4);
    const wrapper = render(
      <HistoryDisplay
        history={h}
        names={['', '', '', '']}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders basic history consistently', () => {
    const h = new PlayerHistory(4);
    h.add(1);
    h.add(2);
    h.newGroup();
    h.add(0);
    h.add(0);
    h.newGroup();
    h.add(3);
    const wrapper = render(
      <HistoryDisplay
        history={h}
        names={['', '', '', '']}
      />
    );
    expect(wrapper).toMatchSnapshot();
  })
});
