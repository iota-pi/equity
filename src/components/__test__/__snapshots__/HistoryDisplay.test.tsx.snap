// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HistoryDisplay/> renders base state consistently 1`] = `
<div
  class="HistoryDisplay-root-1"
>
  <div
    class="HistoryDisplay-scrollable-2"
    style="margin-right:0"
  >
    <div
      class="HistoryDisplay-historyContainer-3"
    >
      <div
        style="position:relative"
      >
        <div
          class="HistoryDisplay-verticalSpacer-5"
        />
        <div
          class="HistoryDisplay-verticalSpacer-5"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<HistoryDisplay/> renders basic history consistently 1`] = `
<div
  class="HistoryDisplay-root-1"
>
  <div
    class="HistoryDisplay-scrollable-2"
    style="margin-right:0"
  >
    <div
      class="HistoryDisplay-historyContainer-3"
    >
      <div
        style="position:relative"
      >
        <div
          class="HistoryDisplay-verticalSpacer-5"
        />
        <h1
          class="MuiTypography-root MuiTypography-h1 MuiTypography-alignCenter HistoryDisplay-historyRow-4"
        >
          <span>
            2
          </span>
          <span>
            , 
          </span>
          <span>
            3
          </span>
        </h1>
        <h1
          class="MuiTypography-root MuiTypography-h1 MuiTypography-alignCenter HistoryDisplay-historyRow-4"
        >
          <span>
            1
          </span>
        </h1>
        <h1
          class="MuiTypography-root MuiTypography-h1 MuiTypography-alignCenter HistoryDisplay-historyRow-4"
        >
          <span>
            1
          </span>
        </h1>
        <h1
          class="MuiTypography-root MuiTypography-h1 MuiTypography-alignCenter HistoryDisplay-historyRow-4"
        >
          <span>
            4
          </span>
        </h1>
        <div
          class="HistoryDisplay-verticalSpacer-5"
        />
      </div>
    </div>
  </div>
</div>
`;
